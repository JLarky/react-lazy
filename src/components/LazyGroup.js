import React from 'react'
import PropTypes from 'prop-types'

import { countTypesTags, wrapTypesToLazyChild, wrapTypesToNoScript } from '../lib/wrap'

import DefaultWrapper from './DefaultWrapper'
import Observer from '@researchgate/react-intersection-observer'

class LazyGroup extends React.PureComponent {
    constructor(props) {
        super(props)

        this.loadedImgTags = 0
        this.state = { imgTagCount: null, loadedAt: null, viewportAt: null }

        this.handleViewport = this.handleViewport.bind(this)
        this.handleImgLoaded = this.handleImgLoaded.bind(this)
    }

    handleImgLoaded() {
        this.loadedImgTags++

        if (this.loadedImgTags === this.state.imgTagCount) {
            this.loadedImgTags = 0
            this.setState({ imgTagCount: null, loadedAt: Date.now() }, this.props.onLoad)
        }
    }

    handleViewport(event, unobserve) {
        const { children, childrenToWrap, onLoad, onViewport, visible } = this.props

        if (!event.isIntersecting || !visible) {
            return
        }

        if (onViewport) {
            onViewport(event)
        }

        if (event.defaultPrevented) {
            return
        }

        unobserve()

        const imgTagCount = countTypesTags(childrenToWrap, children) || null
        this.loadedImgTags = 0
        const viewportAt = Date.now()
        this.setState(
            { imgTagCount, loadedAt: !imgTagCount ? viewportAt : null, viewportAt },
            !imgTagCount ? onLoad : null,
        )
    }

    render() {
        const {
            children,
            childrenToWrap,
            childWrapper,
            clientOnly,
            component,
            cushion,
            forwardedRef: ref,
            ltIE9,
            onLoad,
            onViewport,
            threshold,
            viewport,
            visible,
            ...rest
        } = this.props

        const props = Object.assign({ ref }, rest)

        return (
            <Observer onChange={this.handleViewport} root={viewport} rootMargin={cushion} threshold={threshold}>
                {React.createElement(
                    component,
                    props,
                    // swap render once element is visible in viewport
                    clientOnly || this.state.viewportAt
                        ? wrapTypesToLazyChild(childrenToWrap, children, childWrapper, this.handleImgLoaded, this.state.viewportAt != null)
                        : wrapTypesToNoScript(childrenToWrap, children, ltIE9, childWrapper),
                )}
            </Observer>
        )
    }
}

LazyGroup.defaultProps = {
    childrenToWrap: ['iframe', 'img'],
    childWrapper: DefaultWrapper,
    clientOnly: false,
    component: 'div',
    ltIE9: false,
    visible: true,
}

LazyGroup.propTypes = {
    clientOnly: PropTypes.bool,
    children: PropTypes.node,
    childrenToWrap: PropTypes.arrayOf(PropTypes.any),
    childWrapper: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    component: PropTypes.any,
    cushion: PropTypes.string,
    forwardedRef: PropTypes.any,
    ltIE9: PropTypes.bool,
    onLoad: PropTypes.func,
    onViewport: PropTypes.func,
    threshold: PropTypes.oneOfType([PropTypes.number, PropTypes.arrayOf(PropTypes.number)]),
    viewport: PropTypes.oneOfType(
        [PropTypes.string].concat(typeof HTMLElement === 'undefined' ? [] : PropTypes.instanceOf(HTMLElement)),
    ),
    visible: PropTypes.bool,
}

// eslint-disable-next-line react/display-name,react/no-multi-comp
export default React.forwardRef((props, ref) => <LazyGroup {...props} forwardedRef={ref} />)
