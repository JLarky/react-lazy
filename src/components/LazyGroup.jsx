import React from 'react'
import { findDOMNode } from 'react-dom'
import PropTypes from 'prop-types'

import { addElement, removeElement } from '../lib/viewport'
import { countTypesTags, wrapTypesToLazyChild, wrapTypesToNoScript } from '../lib/wrap'

class LazyGroup extends React.PureComponent {
    constructor(props) {
        super(props)

        this.loadedImgTags = 0
        this.state = { imgTagCount: null, loadedAt: null, viewportAt: null }

        this.onImgLoaded = this.onImgLoaded.bind(this)
        this.onViewport = this.onViewport.bind(this)
    }

    componentDidMount() {
        this.options = {
            callback: this.onViewport,
            cushion: this.props.cushion,
            element: findDOMNode(this),
        }
        addElement(this.options)
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.cushion !== this.props.cushion) {
            this.options.cushion = nextProps.cushion
        }
    }

    componentWillUnmount() {
        removeElement(this.options)
        delete this.options
    }

    onImgLoaded() {
        this.loadedImgTags++

        if (this.loadedImgTags === this.state.imgTagCount) {
            this.loadedImgTags = 0
            this.setState({ imgTagCount: null, loadedAt: Date.now() }, this.props.onLoad)
        }
    }

    onViewport() {
        const { children, childrenToWrap, onLoad, onViewport, visible } = this.props

        if (!visible) {
            return false
        }

        const imgTagCount = countTypesTags(childrenToWrap, children) || null
        this.loadedImgTags = 0
        if (onViewport) {
            onViewport()
        }
        const viewportAt = Date.now()
        this.setState(
            { imgTagCount, loadedAt: !imgTagCount ? viewportAt : null, viewportAt },
            !imgTagCount ? onLoad : null
        )
    }

    render() {
        const { children, childrenToWrap, childWrapper, component, ltIE9, visible, ...props } = this.props
        // remove props we do not want to pass to the rendering component
        delete props.cushion
        delete props.onLoad
        delete props.onViewport

        return React.createElement(
            component,
            props,
            // swap render once element is visible in viewport
            visible && this.state.viewportAt
            // replace elements with LazyChild
            ? wrapTypesToLazyChild(childrenToWrap, children, childWrapper, this.onImgLoaded)
            // wrap given element types to noscript and the given wrapper component
            : wrapTypesToNoScript(childrenToWrap, children, ltIE9, childWrapper)
        )
    }
}

LazyGroup.defaultProps = {
    component: 'div',
    childrenToWrap: ['iframe', 'img'],
    cushion: 0,
    ltIE9: false,
    visible: true,
}

const ReactElement = PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.func,
])

LazyGroup.propTypes = {
    children: PropTypes.node,
    childrenToWrap: PropTypes.arrayOf(ReactElement),
    childWrapper: ReactElement,
    component: ReactElement,
    cushion: PropTypes.number,
    ltIE9: PropTypes.bool,
    onLoad: PropTypes.func,
    onViewport: PropTypes.func,
    visible: PropTypes.bool,
}

export default LazyGroup
