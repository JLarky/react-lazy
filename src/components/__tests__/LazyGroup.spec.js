/* eslint-env jest */
/* eslint-disable react/prop-types,react/no-multi-comp */
import React from 'react'
import renderer from 'react-test-renderer'

import LazyGroup from '../LazyGroup'

const originalIntersectionObserver = window.IntersectionObserver

function noop() {}

function mockIntersectionObserver() {
    return {
        disconnect: noop,
        observe: noop,
        unobserve: noop,
    }
}

const target = { nodeType: 1 }

beforeAll(() => {
    window.IntersectionObserver = mockIntersectionObserver
})

afterAll(() => {
    window.IntersectionObserver = originalIntersectionObserver
})

describe('LazyGroup', () => {
    test('should render just a div by default', () => {
        const tree = renderer.create(<LazyGroup />)
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('should let change component to be rendered', () => {
        const tree = renderer.create(<LazyGroup component="section" />)
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('ltIE9 should render just a div by default', () => {
        const tree = renderer.create(<LazyGroup ltIE9 />)
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('should render span inside div by default', () => {
        const tree = renderer.create(
            <LazyGroup>
                <span>I am NOT inside noscript!</span>
            </LazyGroup>,
        )
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('childrenToWrap should wrap given component types inside noscript', () => {
        const tree = renderer.create(
            <LazyGroup childrenToWrap={['span']}>
                <span>I am inside noscript!</span>
                <div>
                    I am not! <span>But I am!</span>
                </div>
            </LazyGroup>,
        )
        expect(tree.toJSON()).toMatchSnapshot()

        const tree2 = renderer.create(
            <LazyGroup childrenToWrap={['span']} ltIE9>
                <span>I am inside IECC noscript!</span>
                <div>
                    I am not! <span>But I am!</span>
                </div>
            </LazyGroup>,
        )
        expect(tree2.toJSON()).toMatchSnapshot()
    })

    test('ltIE9 should render span inside div by default', () => {
        const tree = renderer.create(
            <LazyGroup ltIE9>
                <span>I am NOT inside IECC noscript!</span>
            </LazyGroup>,
        )
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('childrenToWrap should wrap matching children in clientOnly mode', () => {
        const tree = renderer.create(
            <LazyGroup clientOnly>
                <img alt="I should be replaced by a placeholder!" />
                <div>
                    Not wrapped! You can read me in the snapshot! <img alt="Replaced by placeholder!" />
                </div>
            </LazyGroup>,
        )
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('should call ref callback', () => {
        const spy = jest.fn()
        renderer.create(<LazyGroup ref={spy} />, { createNodeMock: () => target })
        expect(spy).toHaveBeenCalledWith(target)
    })
})
