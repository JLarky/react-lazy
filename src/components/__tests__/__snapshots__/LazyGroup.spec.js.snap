// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyGroup childrenToWrap should wrap given component types inside noscript 1`] = `
<div>
  <div
    className="react-lazy-wrapper react-lazy-wrapper--placeholder"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<noscript><span>I am inside noscript!</span></noscript>",
      }
    }
  />
  <div>
    I am not! 
    <div
      className="react-lazy-wrapper react-lazy-wrapper--placeholder"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<noscript><span>But I am!</span></noscript>",
        }
      }
    />
  </div>
</div>
`;

exports[`LazyGroup childrenToWrap should wrap given component types inside noscript 2`] = `
<div>
  <div
    className="react-lazy-wrapper react-lazy-wrapper--placeholder"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<!--[if IE 9]><!--><noscript><!--<![endif]--><span>I am inside IECC noscript!</span><!--[if IE 9]><!--></noscript><!--<![endif]-->",
      }
    }
  />
  <div>
    I am not! 
    <div
      className="react-lazy-wrapper react-lazy-wrapper--placeholder"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<!--[if IE 9]><!--><noscript><!--<![endif]--><span>But I am!</span><!--[if IE 9]><!--></noscript><!--<![endif]-->",
        }
      }
    />
  </div>
</div>
`;

exports[`LazyGroup childrenToWrap should wrap matching children in clientOnly mode 1`] = `
<div>
  <div
    className="react-lazy-wrapper react-lazy-wrapper--placeholder"
  />
  <div>
    Not wrapped! You can read me in the snapshot! 
    <div
      className="react-lazy-wrapper react-lazy-wrapper--placeholder"
    />
  </div>
</div>
`;

exports[`LazyGroup ltIE9 should render just a div by default 1`] = `<div />`;

exports[`LazyGroup ltIE9 should render span inside div by default 1`] = `
<div>
  <span>
    I am NOT inside IECC noscript!
  </span>
</div>
`;

exports[`LazyGroup should let change component to be rendered 1`] = `<section />`;

exports[`LazyGroup should render just a div by default 1`] = `<div />`;

exports[`LazyGroup should render span inside div by default 1`] = `
<div>
  <span>
    I am NOT inside noscript!
  </span>
</div>
`;
